app/src/main/java/com/group80/uoftinder/GenerateCompatibilityList.java

```diff
@@ -0,0 +1,113 @@
+        usf = new UserScoreFacade(curUser, new CreateAccountInteractor()); // may need to change 2nd param
+        User curUser = UserRealtimeDbFacade.getCurrentUser();
+        int curUserScore = curUser.getScore();
+        this.recPresenterInterface = recPresenterInterface;
+    }
+
+    /**
+     * Generate the compatibility list for curUser and update the User's compatibilityList attribute.
+     * Pass the compatibilityList to recPresenterInterface so that it can display it
+     */
+    public void generateCompatibilityList() {
+        compatibilityList = new ArrayList<>();
+        Map<String, Integer> compScores = calculateCompatibilityScores(allUsers);
+        for (int i = 0 ; i < users.size() ; i++) {
+            String maxKey = getMaxKey(compScores);
+            compatibilityList.add(maxKey);
+            compScores.remove(maxKey);
+        }
+//        curUser.setCompatibilityList(compatibilityList);
+//        recPresenterInterface.showUsers(compatibilityList);
+    }
+
+    /**
+     * Return the key that corresponds to the max value in userScore
+     * @param userScore: a map that maps the user ID to their compatibility score with the current
+     *                 user
+     * @return the key that corresponds to the max value in userScore
+     */
+    private String getMaxKey(Map<String, Integer> userScore) {
+        // TODO: change this to get key of max value
+        int maxValue = 0;
+        String maxKey = "";
+        for (String key : userScore.keySet()) {
+            Integer currentScore = userScore.get(key);
+            if (currentScore != null && currentScore > maxValue) {
+//                if (currentScore > maxValue) {
+                maxValue = currentScore;
+                maxKey = key;
+//                }
+            }
+
+        }
+        return maxKey;
+    }
+
+    /**
+     * Return a map of user ID to the user's compatibility score with the current user
+     * @param users: a list of all User in the database
+     * @return a map of user ID to the user's compatibility score with the current user
+     */
+    private Map<String, Integer> calculateCompatibilityScores(List<User> users) {
+        Map<String, Integer> compScores = new Hashtable<>();
+        for (User user : users) {
+            calculateCompatibiltyScore(compScores, user);
+        }
+        return compScores;
+    }
+
+    /**
+     * Mutate compScores to contain a map of compUser's user ID to their compatibility score with
+     * curUser
+     * @param compScores: a map of user IDs to their compatibility score with curUser
+     * @param compUser: the user to compare with curUser
+     */
+    private void calculateCompatibiltyScore(Map<String, Integer> compScores, User compUser) {
+        int userScore = compUser.getScore();
+        int compScore = usf.compare(curUserScore, userScore);
+        compScores.put(compUser.getUid(), compScore);
```

Comment:
Would it make more sense to map User objects directly to compScore instead of the String user ID? This would save us a call to the database in `update` method below to get the User object by user ID and seems more efficient in my opinion? Love to hear more of your thoughts about this that I might have not considered

Reply:
I implemented it this way as I talked with Lance and he said that it would be better (and easier) to map a user id string to their compScore instead of a User object itself to their compScore.

Reply:
I think that you make a valid point, Raghav. I believe that the value in the HashMap is no longer accessible if the HashMap key is mutated. This would be fine if the keys were of type User, as the address of the User object would not change. However, the keys of the HashMap is the User ID, which I believe can be changed (since it's the username). Thus, I believe that in the User Entity class, the method that handles the user changing their username should also call a method (that I will write) that will reassign that key-value pair to ensure the data is preserved. To adhere to clean architecture, an interface may also need to be created in the entity layer that the GenerateCompatibilityList use case will implement.

Reply:
Hmm maybe it would make more sense to make the keys of the HashMap have type User instead then? This would avoid the need to create another method to re-assign the key-value pair, right? I think this would simplify the code in general as well.